MIPS/mip-zkapps.md

```diff
@@ -0,0 +1,960 @@
+see on other blockchains. For example, you can check arbitrary depth subtrees
+with recursive proofs and sums up all balances and checks that all the balances
+sum to zero in order to allow any token interactions that don't create new
+tokens. Using the parenting system, you can also implement complex interactions
+like liquidity tokens in a DEX.
+
+Note that this structure enables extremely powerful usages of tokens, but it
+will behoove the Mina ecosystem to agree on more restrictive token standards so
+built on top of these systems such that groups of contracts can compose with
+one-another more easily.
+
+#### Call Forest Encoding
+
+We can construct a call forest from a list of account updates, where the
+call-depth field of an account update encodes the Node-Left-Right depth-first
+traversal order of the update within a tree. A new tree within a forest is
+indicated by a call-depth of 0. Each successive account update can either
+increase by 1, stay the same, or decrease by an arbitrary amount. Call-depth 0
+defines a new tree in the forest, and a non-zero number defines a child of the
+most recently defined tree.
+
+Here is an example of a forest constructed from a list of account updates with
+call-depths `0 0 1 2 3 1 0` :
+
+```
+ *  *     *
+    | \
+    *  *
+    |
+    *
+    |
+    *
+```
+
+#### Updates in Account Updates
+
+An account update can contain a request to update certain fields in an account.
+Those fields are:
+
+- `app_state`
+- `delegate`
+- `verification_key`
+- `permissions`
+- `zkapp_uri`
+- `token_symbol`
+- `timing`
+- `voting_for`
+
+Note that some of these items are not in the in the `zkapp` part of an account
+(for example, the `delegate` field). A given account update may request to
+update some or all of these items.
+
+The `permissions` field of accounts is expanded to include permissions to update
+each of these items. Besides requiring permission, the `timing` field of an
+account can be set only if the account is currently untimed. The complete set of
+permissions is described below.
+
+#### Preconditions in account updates
+
+There are three kinds of preconditions that can be checked before applying an
+account update:
+
+- account preconditions
+- network preconditions
+- a `validWhile` precondition
+
+The account preconditions are:
+
+- `balance` : a range of account balances
+- `nonce` : a specific nonce, or a range of nonces
+- `receipt_chain_hash` : a specific receipt chain hash
+- `delegate` : a specific public key
+- `state` : a specific application state
+- `sequence_state` : a specific sequence state
+- `proved_state` : a specific proved state
+- `is_new` : whether the account is new
+
+An account update can include some or all of these
+preconditions. These preconditions are checked against the account
+from the previous protocol state. These conditions must be true at the
+moment the transaction is included in a block, otherwise the
+transaction will fail.
+
+The network preconditions are:
```

Comment:
wow I can't believe we missed that -- however, I don't think it's urgent to get it in before fixing this:

> The initial implementation proposes that network preconditions are to be true at
> the moment the transaction is included in a block, otherwise the transaction
> fails. This failure is known to be a bit frustrating as many of the network
> preconditions change rapidly. In a future hardfork, there will likely be changes
> to allow network preconditions to be true in a range of slots/blocks rather than
> the exact one where the transaction is added.

The staged ledger hash changes every block so it will be effectively unusable initially.

Reply:
Hard to add, I guess? Yeah I think could be a later thing if so

Reply:
Not hard to add once the blocking change above is ready, but that one is a bit risky

Reply:
Gotcha, okay.